### PR TITLE
tracing-attributes: support const values for target and name

### PR DESCRIPTION
**Pull request by dpc**
Fixes: https://github.com/tokio-rs/tracing/issues/2460

I only fixed  lints making #2917 fail. Seems like they affect `master` branch as well?

### CODE REVIEW

File: tracing-attributes/src/attr.rs [hunk 1]
```diff
@@ -86,6 +86,8 @@ impl Parse for InstrumentArgs {
                 // XXX: apparently we support names as either named args with an
                 // sign, _or_ as unnamed string literals. That's weird, but
                 // changing it is apparently breaking.
+                // This also means that when using idents for name, it must be via
+                // a named arg.
```
Comment (hds) [inline tracing-attributes/src/attr.rs:90]: To make this comment a bit clearer for the next person who is reading it, it would be great if you included an example of what you mean.

File: tracing-attributes/tests/instrument.rs [hunk 2]
```diff
@@ -238,3 +239,38 @@ fn impl_trait_return_type() {
 
     handle.assert_finished();
 }
+
+#[test]
+fn target_name_ident() {
```
Comment (hds) [inline tracing-attributes/tests/instrument.rs:244]: Could we split this into 2 or 3 tests please? One for `name`, another for `target` and if you believe it's necessary then a 3rd one for both (I'll leave that up to you).

File: tracing-core/src/dispatch.rs [hunk 3]
```diff
@@ -510,6 +510,7 @@ pub(crate) fn get_global() -> &'static Dispatch {
     unsafe {
         // This is safe given the invariant that setting the global dispatcher
         // also sets `GLOBAL_INIT` to `INITIALIZED`.
+        #[allow(static_mut_refs)]
```
Comment (hds) [inline tracing-core/src/dispatch.rs:513]: We'd prefer to have these changes to adhere to new clippy lints split into a separate PR to go in first. This keeps the clippy changes isolated from the actual fix.

On the up side, you get 2 PRs on `tracing` for the price of 1!

**Review by hds — DISMISSED** [review]
Thanks for your PR!

There are a couple of comments I've added regarding test layout and the code comment. Also, it would be best to separate the changes to make clippy pass into a separate PR (e.g. #2814).

File: tracing-core/src/dispatch.rs (hunk 3 above)
Comment (dpc) [reply tracing-core/src/dispatch.rs:513]: #2948

File: tracing-attributes/tests/instrument.rs (hunk 2 above)
Comment (dpc) [reply tracing-attributes/tests/instrument.rs]: Done.

File: tracing-attributes/src/attr.rs (hunk 1 above)
Comment (dpc) [reply tracing-attributes/src/attr.rs:90]: I'm sorry but I don't understand myself. I have some vague idea, but that's about it.

I just really need this feature, and I noticed that there's already a PR by someone else lingering, so I just picked it up, refreshed and making sure it lands.

Comment (dpc) [reply tracing-attributes/src/attr.rs:90]: Oh, looks like the original author @mbrobbel is around. If anyone can paste the string I'm supposed to add there, I'm happy to do so and push the result. :D

Comment (mbrobbel) [reply tracing-attributes/src/attr.rs:90]: It uses a different parser when there is no named arg (`name =`), so you have to use `name = ` if you want to use an identifier of a const value for name.

Comment (dpc) [reply tracing-attributes/src/attr.rs:90]:
```
#[instrument(name = SOME_IDENT)]
```

instead of 

```
#[instrument(SOME_IDENT)]
```

?

**Review by mbrobbel — APPROVED** [review]

**Comment by dpc** [comment]
I miss this feature so much, every day. :D

What's the best way to give it another review, approval, merge, release? :D

**Review by hawkw — APPROVED** [review]
This looks good to me, thanks for addressing the feedback from @hds!
